src/messaging/ExchangeContext.cpp

```diff
@@ -308,25 +370,80 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
     // layer has completed its work on the ExchangeContext.
     Retain();
 
+    messageId   = packetHeader.GetMessageId();
     protocolId  = payloadHeader.GetProtocolID();
     messageType = payloadHeader.GetMessageType();
 
-    // Since we got the response, cancel the response timer.
-    CancelResponseTimer();
+    if (payloadHeader.IsAckMsg())
+    {
+        Optional<uint32_t> ackId = payloadHeader.GetAckId();
 
-    // If the context was expecting a response to a previously sent message, this message
-    // is implicitly that response.
-    SetResponseExpected(false);
+        if (ackId.HasValue())
+        {
+            err = mReliableMessageContext.HandleRcvdAck(ackId.Value());
+        }
+        else
+        {
+            err = CHIP_ERROR_DECODE_FAILED;
+            ChipLogError(ExchangeManager, "Failed to retrieve Ack ID");
+            ExitNow(err = CHIP_NO_ERROR);
```

Comment:
This part really doesn't make sense; then why bother setting to `CHIP_ERROR_DECODE_FAILED` above?